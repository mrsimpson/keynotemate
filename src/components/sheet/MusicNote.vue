<template>
  <div
    class="music-note"
    :style="{
      bottom: `${calculatePosition()}px`,
      left: `${index * 40}px`,
    }"
  >
    <div
      class="note-circle"
      :style="{
        backgroundColor: getNoteColor(pitch),
      }"
    >
      <span class="note-label">{{ pitch }}</span>
    </div>
  </div>
</template>

<script setup lang="ts">
const { getNoteColor } = useNoteVisualizer();

const props = defineProps<{
  pitch: string;
  octave: number;
  index: number;
}>();

const notePositions: Record<string, number> = {
  C: 0,
  D: 4,
  E: 8,
  F: 12,
  G: 16,
  A: 20,
  H: 24,
};

const calculatePosition = () => {
  const basePosition = notePositions[props.pitch] || 0;
  const octaveOffset = (props.octave - 4) * 28; // 28px per octave
  return basePosition + octaveOffset;
};
</script>

<style scoped>
.music-note {
  position: absolute;
  width: 30px;
  height: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.note-circle {
  width: 24px;
  height: 24px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
}

.note-label {
  color: white;
  font-size: 12px;
  font-weight: bold;
  text-shadow: 1px 1px 1px rgba(0, 0, 0, 0.3);
}
</style>
