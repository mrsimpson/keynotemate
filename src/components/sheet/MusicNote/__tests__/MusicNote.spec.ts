import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import MusicNote from '@/components/sheet/MusicNote.vue'

describe('MusicNote Component', () => {
  it('renders music note snapshot', () => {
    const wrapper = mount(MusicNote, {
      props: {
        pitch: 'C',
        octave: 4,
        index: 0,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders music note with different pitch snapshot', () => {
    const wrapper = mount(MusicNote, {
      props: {
        pitch: 'G',
        octave: 5,
        index: 1,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
