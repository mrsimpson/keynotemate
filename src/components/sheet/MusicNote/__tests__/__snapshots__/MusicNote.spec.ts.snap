// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MusicNote Component > renders music note snapshot 1`] = `
"<div data-v-74367537="" class="music-note" style="bottom: 0px; left: 0px;">
  <div data-v-74367537="" class="note-circle" style="background-color: rgb(255, 0, 0);"><span data-v-74367537="" class="note-label">C</span></div>
</div>"
`;

exports[`MusicNote Component > renders music note with different pitch snapshot 1`] = `
"<div data-v-74367537="" class="music-note" style="bottom: 44px; left: 40px;">
  <div data-v-74367537="" class="note-circle" style="background-color: rgb(0, 0, 255);"><span data-v-74367537="" class="note-label">G</span></div>
</div>"
`;
